package cetty.craft;

option java_package = "io.cetty.craft";
option java_outer_classname = "CraftProto";

enum Lang {
	(craft_enum_options).query_params="l";
	
	en_US = 0;
    zh_CN = 1;
    zh_TW = 2;
}

message KeyValue {
	(craft_message_options).code = "{key}-{value}";
	
	optional string   key = 1;
	optional string value = 2;
}

message Status {
	message DetailError {
		optional string    code = 1;
		optional string message = 2;
	}
	
	optional int32   http_status = 1 [default = 200];
    optional int32          code = 2 [default = 0];
    optional string      message = 3;
	optional string    more_info = 4;
	
	repeated DetailError  errors = 9;
}

message Session {
	optional string   id = 1[
		(craft).cookie_param = "id",
	];
	
	optional string user = 2;
}

message Authority {
	optional Session  session = 1;
	
    optional string       key = 2;
	
    optional string signature = 3 [
		(craft).query_param = "",
	];

	optional string        ip = 8;
	optional string      host = 9; // non-ip
}

message ErrorResponse {
	optional Status status = 1;
}

message CraftFieldOptions {
    optional string   path_param = 1;
	optional string  query_param = 2;
	optional string header_param = 3;
	optional string cookie_param = 4;
	
	optional string      default = 5;
	
	optional bool    ignore_sign = 20;
	optional string      pattern = 21;
	optional int32   decimal_min = 22;
	optional int32   decimal_max = 23;
}

message CraftMessageOptions {
	optional string query_param = 2;
}

message CraftMethodOptions {
	optional string method = 1;
	optional string   path = 2;
	optional string  query = 3;
	optional string header = 4;
	optional string cookie = 5;
	
	optional int32  timeout = 10;
}

message CraftServiceOptions {
	optional string   path = 1;
}

extend google.protobuf.FieldOptions {
  optional CraftFieldOptions craft = 60100;
}

extend google.protobuf.MessageOptions {
	optional CraftMessageOptions craft_message_options = 60100;
}

extend google.protobuf.MethodOptions {
  optional CraftMethodOptions craft_method_options = 60100;
}

extend google.protobuf.ServiceOptions {
  optional CraftServiceOptions craft_service_options = 60100;
}
